diff --git a/proof/invariant-abstract/AARCH64/ArchInvariants_AI.thy b/proof/invariant-abstract/AARCH64/ArchInvariants_AI.thy
index bea5c788..f9111422 100644
--- a/proof/invariant-abstract/AARCH64/ArchInvariants_AI.thy
+++ b/proof/invariant-abstract/AARCH64/ArchInvariants_AI.thy
@@ -1494,10 +1494,9 @@ lemma canonical_user_ge0[intro!,simp]:
   "0 < canonical_user"
   by (simp add: canonical_user_def mask_def ipa_size_def)
 
-lemma pptr_base_kernel_elf_base: (* FIXME MK: candidate for Kernel_Config_Lemmas *)
+lemma pptr_base_kernel_elf_base:
   "pptr_base < kernel_elf_base"
-  by (simp add: pptr_base_def pptrBase_def canonical_bit_def kernel_elf_base_def kernelELFBase_def
-                kernelELFPAddrBase_def Kernel_Config.physBase_def mask_def)
+  by (simp add: pptr_base_def kernel_elf_base_def pptrBase_kernelELFBase)
 
 lemma pptrTop_le_ipa_size:
   "pptrTop \<le> mask ipa_size"
diff --git a/proof/invariant-abstract/AARCH64/ArchKernelInit_AI.thy b/proof/invariant-abstract/AARCH64/ArchKernelInit_AI.thy
index 719c4cf6..2edbd30d 100644
--- a/proof/invariant-abstract/AARCH64/ArchKernelInit_AI.thy
+++ b/proof/invariant-abstract/AARCH64/ArchKernelInit_AI.thy
@@ -248,15 +248,6 @@ lemma idle_thread_in_kernel_window_init_arch_state[simp]:
   apply (rule conjI; unat_arith)
   done
 
-lemma irq_node_pptr_base_kernel_elf_base:
-  "\<lbrakk>x \<le> pptr_base + (m + (mask cte_level_bits + 0x3000)); m \<le> mask (size irq) << cte_level_bits \<rbrakk>
-   \<Longrightarrow> \<not> kernel_elf_base \<le> x" for irq::irq
-  apply (simp add: word_size cte_level_bits_def mask_def pptr_base_def pptrBase_def
-                   Kernel_Config.physBase_def kernel_elf_base_def kernelELFBase_def canonical_bit_def
-                   not_le)
-  apply unat_arith
-  done
-
 lemma irq_node_in_kernel_window_init_arch_state':
   "\<lbrakk> init_irq_node_ptr + m \<le> x; x \<le> init_irq_node_ptr + m + mask cte_level_bits;
      m \<le> mask (size (irq::irq)) << cte_level_bits\<rbrakk>
