diff --git a/proof/invariant-abstract/AARCH64/Machine_AI.thy b/proof/invariant-abstract/AARCH64/Machine_AI.thy
index 91b4ed83..56ee5eda 100644
--- a/proof/invariant-abstract/AARCH64/Machine_AI.thy
+++ b/proof/invariant-abstract/AARCH64/Machine_AI.thy
@@ -291,7 +291,6 @@ crunch_ignore (valid, empty_fail, no_fail)
    grep -oE "(\w+_impl)|(get\w+)" MachineOps.thy|sort|uniq|sed "s/_impl//;s/$/,/;s/^/  /"
    with the following manual interventions:
    - remove false positives: get_def, gets_def, getFPUState, getRegister, getRestartPC
-   - add readVCPUHardwareReg (which uses non-standard "Val" instead of "_val" (FIXME AARCH64))
    - add read_cntpct
    - remove final comma
    - getActiveIRQ does not preserve no_irq *)
