diff --git a/proof/invariant-abstract/AARCH64/ArchAInvsPre.thy b/proof/invariant-abstract/AARCH64/ArchAInvsPre.thy
index 809c5e80..5f77c0be 100644
--- a/proof/invariant-abstract/AARCH64/ArchAInvsPre.thy
+++ b/proof/invariant-abstract/AARCH64/ArchAInvsPre.thy
@@ -42,27 +42,6 @@ lemma data_at_aligned:
   "\<lbrakk> data_at sz p s; pspace_aligned s\<rbrakk> \<Longrightarrow> is_aligned p (pageBitsForSize sz)"
   unfolding data_at_def by (auto simp: obj_at_def dest: pspace_alignedD)
 
-lemma is_aligned_pptrBaseOffset_pt_bits_left:
-  "level \<le> max_pt_level \<Longrightarrow> is_aligned pptrBaseOffset (pt_bits_left level)"
-  apply (drule pt_bits_left_le_max_pt_level)
-  apply (fastforce elim: is_aligned_weaken[rotated]
-                   simp: pptrBaseOffset_def pptrBase_def paddrBase_def
-                         is_aligned_def pt_bits_left_bound_def bit_simps
-                   split: if_split_asm)
-  done
-
-lemma is_aligned_ptrFromPAddr_n_eq:
-  "level \<le> max_pt_level \<Longrightarrow>
-   is_aligned (ptrFromPAddr x) (pt_bits_left level) = is_aligned x (pt_bits_left level)"
-  apply (rule iffI)
-   apply (simp add: ptrFromPAddr_def)
-   apply (erule is_aligned_addD2)
-   apply (erule is_aligned_pptrBaseOffset_pt_bits_left)
-  apply (erule is_aligned_ptrFromPAddr_n)
-  apply (drule pt_bits_left_le_max_pt_level)
-  apply (clarsimp simp: pt_bits_left_bound_def bit_simps split: if_split_asm)
-  done
-
 lemma some_get_page_info_umapsD:
   "\<lbrakk>get_page_info (aobjs_of s) pt_ref vptr = Some (b, a, attr, r);
     \<exists>\<rhd> (max_pt_level, pt_ref) s; vptr \<in> user_region; valid_vspace_objs s; pspace_aligned s;
diff --git a/proof/invariant-abstract/AARCH64/ArchAcc_AI.thy b/proof/invariant-abstract/AARCH64/ArchAcc_AI.thy
index a3ab073e..adcdb299 100644
--- a/proof/invariant-abstract/AARCH64/ArchAcc_AI.thy
+++ b/proof/invariant-abstract/AARCH64/ArchAcc_AI.thy
@@ -206,17 +206,6 @@ lemma unique_vs_lookup_table:
   apply (drule table_cap_ref_vs_cap_ref; simp)
   done
 
-(* FIXME AARCH64: move *)
-lemma level_type_less_max_pt_level:
-  "level < max_pt_level \<Longrightarrow> level_type level = NormalPT_T"
-  by (clarsimp simp: level_type_def)
-
-(* FIXME AARCH64: move *)
-(* Depending on config, ptTranslationBits NormalPT_T might be equal to ptTranslationBits VSRootPT_T *)
-lemma ptTranslationBits_NormalPT_T_leq:
-  "ptTranslationBits NormalPT_T \<le> ptTranslationBits VSRootPT_T"
-  by (simp add: bit_simps)
-
 lemma vref_for_level_pt_index_idem:
   assumes "level' \<le> max_pt_level" and "level'' \<le> level'" and "level < max_pt_level"
   shows "vref_for_level
@@ -1494,24 +1483,6 @@ lemma kernel_regionsI:
   unfolding kernel_regions_def
   by auto
 
-(* FIXME AARCH64: probably remove
-lemma user_region_canonical_pptr_base:
-  "\<lbrakk> p \<notin> user_region; canonical_address p \<rbrakk> \<Longrightarrow> pptr_base \<le> p"
-  using canonical_below_pptr_base_canonical_user word_le_not_less
-  by (auto simp add: user_region_def not_le) *)
-
-(* FIXME AARCH64: probably remove
-lemma kernel_regions_pptr_base:
-  "\<lbrakk> p \<in> kernel_regions s; valid_uses s \<rbrakk> \<Longrightarrow> pptr_base \<le> p"
-  apply (rule user_region_canonical_pptr_base)
-   apply (simp add: valid_uses_def window_defs)
-   apply (erule_tac x=p in allE)
-   apply auto[1]
-  apply (simp add: valid_uses_def window_defs)
-  apply (erule_tac x=p in allE)
-  apply auto[1]
-  done *)
-
 lemma set_pt_valid_global_vspace_mappings[wp]:
   "\<lbrace>\<top>\<rbrace> set_pt p pt \<lbrace>\<lambda>_. valid_global_vspace_mappings\<rbrace>"
   unfolding valid_global_vspace_mappings_def by wp
diff --git a/proof/invariant-abstract/AARCH64/ArchFinalise_AI.thy b/proof/invariant-abstract/AARCH64/ArchFinalise_AI.thy
index 548065c2..ea3e5ed6 100644
--- a/proof/invariant-abstract/AARCH64/ArchFinalise_AI.thy
+++ b/proof/invariant-abstract/AARCH64/ArchFinalise_AI.thy
@@ -1644,10 +1644,6 @@ crunch pred_tcb_at[wp]:
 crunch pred_tcb_at[wp_unsafe]: arch_finalise_cap "pred_tcb_at proj P t"
   (simp: crunch_simps wp: crunch_wps)
 
-(* FIXME AARCH64
-crunch empty[wp]: find_free_hw_asid, store_hw_asid, load_hw_asid, set_vm_root_for_flush, page_table_mapped, invalidate_tlb_by_asid
-  "\<lambda>s. P (obj_at (empty_table {}) word s)" *)
-
 lemma set_vcpu_empty[wp]:
   "\<lbrace>\<lambda>s. P (obj_at (empty_table {}) word s)\<rbrace> set_vcpu p v \<lbrace>\<lambda>_ s. P (obj_at (empty_table {}) word s)\<rbrace>"
   apply (rule set_vcpu.vsobj_at)
diff --git a/proof/invariant-abstract/AARCH64/ArchInvariants_AI.thy b/proof/invariant-abstract/AARCH64/ArchInvariants_AI.thy
index 8d2c28fd..76c657f2 100644
--- a/proof/invariant-abstract/AARCH64/ArchInvariants_AI.thy
+++ b/proof/invariant-abstract/AARCH64/ArchInvariants_AI.thy
@@ -1301,6 +1301,14 @@ lemma level_type_eq[simp]:
   "(level_type level = VSRootPT_T) = (level = max_pt_level)"
   by (simp add: level_type_def)+
 
+lemma level_type_less_max_pt_level:
+  "level < max_pt_level \<Longrightarrow> level_type level = NormalPT_T"
+  by (clarsimp simp: level_type_def)
+
+lemma ptTranslationBits_NormalPT_T_leq:
+  "ptTranslationBits NormalPT_T \<le> ptTranslationBits VSRootPT_T"
+  by (simp add: bit_simps)
+
 lemma valid_vspace_obj_default'[simp]:
   "\<lbrakk> ao_type = VSpaceObj \<Longrightarrow> level = max_pt_level;
      ao_type = PageTableObj \<Longrightarrow> level \<noteq> max_pt_level \<rbrakk> \<Longrightarrow>
@@ -1691,9 +1699,6 @@ lemma vspace_for_asid_SomeI:
   by (clarsimp simp: entry_for_asid_def pool_for_asid_def entry_for_pool_def vspace_for_pool_def
                      vspace_for_asid_def obind_def)
 
-(* FIXME AARCH64: move up *)
-lemmas ptes_of_def = level_pte_of_def
-
 lemma ptes_of_pts_of:
   "ptes_of s pt_t pte_ptr = Some pte \<Longrightarrow>
     \<exists>pt. pts_of s (pte_ptr && ~~mask (pt_bits pt_t)) = Some pt \<and> pt_t = pt_type pt"
@@ -2236,39 +2241,64 @@ lemma aligned_vref_for_level[simp]:
 
 lemmas pt_walk_0[simp] = pt_walk.simps[where level=0, simplified]
 
-(* FIXME AARCH64: the 39 is from the construction of pptrBaseOffset, not sure if we can de-magic that number *)
+(* The current definition of pptrBaseOffset comes out to 2^39. The number below is the maximum
+   alignment this offset allows. We need this to be greater or equal to pt_bits_left max_pt_level *)
+definition pptrBaseOffset_alignment :: nat where
+  "pptrBaseOffset_alignment \<equiv> 39"
+
+(* sanity check for size *)
+lemma pptrBaseOffset_alignment_pt_bits_left[simp, intro!]:
+  "pt_bits_left max_pt_level \<le> pptrBaseOffset_alignment"
+  by (simp add: bit_simps pt_bits_left_def pptrBaseOffset_alignment_def size_max_pt_level)
+
+(* sanity check for alignment *)
+lemma pptrBaseOffset_aligned[simp, intro!]:
+  "is_aligned pptrBaseOffset pptrBaseOffset_alignment"
+  by (simp add: pptrBaseOffset_alignment_def pptrBaseOffset_def pptrBase_def paddrBase_def
+                is_aligned_def)
+
 lemma is_aligned_addrFromPPtr_n:
-  "\<lbrakk> is_aligned p n; n \<le> 39 \<rbrakk> \<Longrightarrow> is_aligned (addrFromPPtr p) n"
+  "\<lbrakk> is_aligned p n; n \<le> pptrBaseOffset_alignment \<rbrakk> \<Longrightarrow> is_aligned (addrFromPPtr p) n"
   apply (simp add: addrFromPPtr_def)
   apply (erule aligned_sub_aligned)
-   apply (simp add: pptrBaseOffset_def pptrBase_def paddrBase_def)
-   apply (erule is_aligned_weaken[rotated])
-   apply (simp add: is_aligned_def)
-  apply simp
+   apply (erule is_aligned_weaken[OF pptrBaseOffset_aligned])
+  apply (simp add: pptrBaseOffset_alignment_def)
   done
 
 lemma is_aligned_addrFromPPtr[intro!]:
   "is_aligned p pageBits \<Longrightarrow> is_aligned (addrFromPPtr p) pageBits"
-  by (simp add: is_aligned_addrFromPPtr_n pageBits_def)
+  by (simp add: is_aligned_addrFromPPtr_n pageBits_def pptrBaseOffset_alignment_def)
 
-(* FIXME AARCH64: the 39 is from the construction of pptrBaseOffset, via is_aligned_addrFromPPtr_n *)
 lemma is_aligned_ptrFromPAddr_n:
-  "\<lbrakk>is_aligned x sz; sz \<le> 39\<rbrakk>
+  "\<lbrakk>is_aligned x sz; sz \<le> pptrBaseOffset_alignment\<rbrakk>
    \<Longrightarrow> is_aligned (ptrFromPAddr x) sz"
-  apply (simp add: ptrFromPAddr_def pptrBaseOffset_def pptrBase_def paddrBase_def)
+  unfolding ptrFromPAddr_def
   apply (erule aligned_add_aligned)
-   apply (erule is_aligned_weaken[rotated])
-   apply (simp add: is_aligned_def)
+   apply (erule is_aligned_weaken[OF pptrBaseOffset_aligned])
   apply (rule order.refl)
   done
 
+lemma is_aligned_pptrBaseOffset_pt_bits_left:
+  "level \<le> max_pt_level \<Longrightarrow> is_aligned pptrBaseOffset (pt_bits_left level)"
+  by (blast intro: order_trans pt_bits_left_mono is_aligned_weaken)
+
+lemma is_aligned_ptrFromPAddr_n_eq:
+  "level \<le> max_pt_level \<Longrightarrow>
+   is_aligned (ptrFromPAddr x) (pt_bits_left level) = is_aligned x (pt_bits_left level)"
+  apply (rule iffI)
+   apply (simp add: ptrFromPAddr_def)
+   apply (erule is_aligned_addD2)
+   apply (erule is_aligned_pptrBaseOffset_pt_bits_left)
+  apply (blast intro: order_trans pt_bits_left_mono is_aligned_ptrFromPAddr_n)
+  done
+
 lemma is_aligned_ptrFromPAddr:
   "is_aligned p pageBits \<Longrightarrow> is_aligned (ptrFromPAddr p) pageBits"
-  by (simp add: is_aligned_ptrFromPAddr_n pageBits_def)
+  by (simp add: is_aligned_ptrFromPAddr_n pageBits_def pptrBaseOffset_alignment_def)
 
 lemma is_aligned_ptrFromPAddr_pt_bits[intro!]:
   "is_aligned p (pt_bits pt_t) \<Longrightarrow> is_aligned (ptrFromPAddr p) (pt_bits pt_t)"
-  by (simp add: is_aligned_ptrFromPAddr_n bit_simps)
+  by (simp add: is_aligned_ptrFromPAddr_n bit_simps pptrBaseOffset_alignment_def)
 
 lemma pspace_aligned_pts_ofD:
   "\<lbrakk> pspace_aligned s; pts_of s pt_ptr = Some pt \<rbrakk> \<Longrightarrow> is_aligned pt_ptr (pt_bits (pt_type pt))"
