diff --git a/proof/invariant-abstract/RISCV64/ArchInvariants_AI.thy b/proof/invariant-abstract/RISCV64/ArchInvariants_AI.thy
index 86892935..1282e5da 100644
--- a/proof/invariant-abstract/RISCV64/ArchInvariants_AI.thy
+++ b/proof/invariant-abstract/RISCV64/ArchInvariants_AI.thy
@@ -1,4 +1,5 @@
 (*
+ * Copyright 2023, Proofcraft Pty Ltd
  * Copyright 2020, Data61, CSIRO (ABN 41 687 119 230)
  *
  * SPDX-License-Identifier: GPL-2.0-only
@@ -1280,10 +1281,55 @@ lemma canonical_user_pptr_base:
   "canonical_user < pptr_base"
   by (simp add: canonical_user_def pptr_base_def pptrBase_def canonical_bit_def mask_def)
 
-lemma pptr_base_kernel_elf_base: (* FIXME MK: candidate for Kernel_Config_Lemmas *)
+(* Shadow lemmas from Arch_Kernel_Config_Lemmas and fold definitions to avoid magic numbers.
+   Written out to make sure the folding has succeeded: *)
+
+lemma is_page_aligned_physBase:
+  "is_aligned physBase pageBits"
+  by (rule is_page_aligned_physBase[folded pageBits_def])
+
+lemma kernel_window_sufficient:
+  "pptrBase + (1 << kernel_window_bits) \<le> kernelELFBase"
+  by (rule kernel_window_sufficient[folded kernel_window_bits_def])
+
+lemma kernel_elf_window_at_least_page:
+  "kernelELFBase + 2 ^ pageBits \<le> kdevBase"
+  by (rule kernel_elf_window_at_least_page[folded pageBits_def])
+
+lemma kernelELFBase_no_overflow:
+  "kernelELFBase < kernelELFBase + 2 ^ pageBits"
+  by (rule kernelELFBase_no_overflow[folded pageBits_def])
+
+(* end shadowing of Arch_Kernel_Config_Lemmas *)
+
+lemma is_page_aligned_kernelELFPAddrBase:
+  "is_aligned kernelELFPAddrBase pageBits"
+  unfolding kernelELFPAddrBase_def
+  by (fastforce intro!: is_aligned_add is_page_aligned_physBase
+                simp: kernelELFPAddrBase_def pageBits_def is_aligned_def)
+
+lemma pptr_base_kernel_elf_base:
   "pptr_base < kernel_elf_base"
-  by (simp add: pptr_base_def pptrBase_def canonical_bit_def kernel_elf_base_def kernelELFBase_def
-                kernelELFPAddrBase_def Kernel_Config.physBase_def mask_def)
+  by (simp add: pptr_base_def kernel_elf_base_def pptrBase_kernelELFBase)
+
+lemma pptr_base_kdev_base:
+  "pptr_base < kdev_base"
+  by (simp add: pptr_base_def pptrBase_def kdev_base_def kdevBase_def canonical_bit_def)
+
+lemma is_page_aligned_pptrTop:
+  "is_aligned pptrTop pageBits"
+  by (simp add: pptrTop_def pageBits_def is_aligned_def)
+
+lemma is_page_aligned_kernel_elf_base:
+  "is_aligned kernel_elf_base pageBits"
+  unfolding kernel_elf_base_def kernelELFBase_def
+  by (simp add: is_aligned_add is_page_aligned_pptrTop is_aligned_andI1
+                is_page_aligned_kernelELFPAddrBase)
+
+lemma canonical_user_kernel_elf_base:
+  "canonical_user < kernel_elf_base"
+  using canonical_user_pptr_base pptr_base_kernel_elf_base
+  by simp
 
 lemma above_pptr_base_canonical:
   "pptr_base \<le> p \<Longrightarrow> canonical_address p"
diff --git a/proof/invariant-abstract/RISCV64/ArchKernelInit_AI.thy b/proof/invariant-abstract/RISCV64/ArchKernelInit_AI.thy
index 3eed9dba..04300d75 100644
--- a/proof/invariant-abstract/RISCV64/ArchKernelInit_AI.thy
+++ b/proof/invariant-abstract/RISCV64/ArchKernelInit_AI.thy
@@ -1,4 +1,5 @@
 (*
+ * Copyright 2023, Proofcraft Pty Ltd
  * Copyright 2020, Data61, CSIRO (ABN 41 687 119 230)
  *
  * SPDX-License-Identifier: GPL-2.0-only
@@ -208,11 +209,33 @@ lemma pt_walk_init_A_st[simp]:
                    is_aligned_pt_slot_offset_pte global_pte_def)
   done
 
+(* Since the bottom 30 bits don't matter for 1GB pages, kernelELFPAddrBase and physBase don't have
+   any influence on the index in the page table, i.e. the virtual address kernel_elf_base is always
+   within the same 1GB page. *)
+lemma elf_index_value:
+  "elf_index = 0x1FE"
+proof -
+  have mask_0: "\<And>n m x. n < m \<Longrightarrow> - (1 << m) && x && mask n = 0"
+    by (metis and.commute and_zero_eq word_mask_and_neg_shift_eq_0 word_bw_assocs(1))
+  have "pt_index max_pt_level kernel_elf_base = 0x1FE"
+    unfolding pt_index_def pt_bits_left_def level_defs bit_simps kernel_elf_base_def
+              kernelELFBase_def pptrTop_def
+    apply (subst word_plus_and_or_coroll)
+     apply (subst mask_0; simp)
+    apply (simp add: word_and_mask_shift_eq_0 shiftr_over_or_dist)
+    apply (simp add: mask_def)
+    done
+  then show ?thesis
+    unfolding elf_index_def
+    by (simp add: bit_simps ucast_ucast_mask)
+qed
+
 lemma table_index_riscv_global_pt_ptr:
   "table_index (pt_slot_offset max_pt_level riscv_global_pt_ptr vref) =
-  ucast ((vref >> ptTranslationBits * 2 + pageBits) && mask ptTranslationBits)"
+  ucast ((vref >> toplevel_bits) && mask ptTranslationBits)"
   apply (simp add: pt_slot_offset_def pt_index_def pt_bits_left_def bit_simps level_defs
-                   riscv_global_pt_ptr_def pptr_base_def pptrBase_def canonical_bit_def)
+                   riscv_global_pt_ptr_def pptr_base_def pptrBase_def canonical_bit_def
+                   toplevel_bits_def)
   apply (subst word_plus_and_or_coroll)
    apply word_bitwise
    apply simp
@@ -220,10 +243,14 @@ lemma table_index_riscv_global_pt_ptr:
   apply (clarsimp simp: word_size)
   done
 
-lemma kernel_window_1G:
-  "\<lbrakk> pptr_base \<le> vref; vref < pptr_base + (1 << 30) \<rbrakk> \<Longrightarrow>
+schematic_goal toplevel_bits_value:
+  "toplevel_bits = ?v"
+  by (simp add: toplevel_bits_def level_defs bit_simps pt_bits_left_def)
+
+lemma kernel_window_bits_table_index:
+  "\<lbrakk> pptr_base \<le> vref; vref < pptr_base + (1 << kernel_window_bits) \<rbrakk> \<Longrightarrow>
     table_index (pt_slot_offset max_pt_level riscv_global_pt_ptr vref) = 0x100"
-  apply (simp add: table_index_riscv_global_pt_ptr)
+  apply (simp add: table_index_riscv_global_pt_ptr toplevel_bits_value kernel_window_bits_def)
   apply (simp add: bit_simps pptr_base_def pptrBase_def neg_mask_le_high_bits word_size flip: NOT_mask)
   apply (subst (asm) mask_def)
   apply (simp add: canonical_bit_def)
@@ -237,81 +264,143 @@ lemma kernel_mapping_slots_0x100[simp]:
                 pptrBase_def canonical_bit_def)
 
 lemma translate_address_kernel_window:
-  "\<lbrakk> pptr_base \<le> vref; vref < pptr_base + (1 << 30) \<rbrakk>\<Longrightarrow>
+  "\<lbrakk> pptr_base \<le> vref; vref < pptr_base + (1 << kernel_window_bits) \<rbrakk>\<Longrightarrow>
    translate_address riscv_global_pt_ptr vref (ptes_of init_A_st) = Some (addrFromPPtr vref)"
   apply (clarsimp simp: translate_address_def in_omonad pt_lookup_target_def
                         pt_lookup_slot_from_level_def)
   apply (simp add: ptes_of_init_A_st_global[THEN fun_cong] init_global_pt_def global_pte_def pte_ref_def)
-  apply (simp add: kernel_window_1G is_aligned_pt_slot_offset_pte)
+  apply (simp add: kernel_window_bits_table_index is_aligned_pt_slot_offset_pte)
   apply (simp add: bit_simps addr_from_ppn_def shiftl_shiftl)
   apply (simp add: ptrFromPAddr_def addrFromPPtr_def)
   apply (simp add: pptrBaseOffset_def paddrBase_def)
-  apply (simp add: pt_bits_left_def bit_simps level_defs)
+  apply (simp add: pt_bits_left_def bit_simps level_defs elf_index_value toplevel_bits_def)
   apply (rule conjI)
    apply (rule is_aligned_add)
     apply (simp add: mask_def)
    apply (simp add: pptrBase_def canonical_bit_def is_aligned_def)
-  apply (simp add: pptr_base_def)
+  apply (simp add: pptr_base_def kernel_window_bits_def)
   apply (simp add: pptrBase_def neg_mask_le_high_bits flip: NOT_mask)
   apply (subst word_plus_and_or_coroll; simp add: canonical_bit_def word_size mask_def)
   apply word_bitwise
   apply clarsimp
   done
 
-lemma elf_window_1M:
-  "\<lbrakk> kernel_elf_base \<le> vref; vref < kernel_elf_base + (1 << 20) \<rbrakk> \<Longrightarrow>
-    table_index (pt_slot_offset max_pt_level riscv_global_pt_ptr vref) = 0x1FE"
-  apply (simp add: table_index_riscv_global_pt_ptr)
-  apply (simp add: bit_simps kernel_elf_base_def kernelELFBase_def kernelELFPAddrBase_def
-                   Kernel_Config.physBase_def)
+lemma kernelELF_plus_page:
+  "((kernelELFPAddrBase && mask toplevel_bits) + 2^pageBits) \<le> 2^toplevel_bits"
+  by (fastforce intro: aligned_mask_plus_bounded is_page_aligned_kernelELFPAddrBase
+                simp: toplevel_bits_value bit_simps)
+
+lemma elf_window_4k:
+  "\<lbrakk> kernel_elf_base \<le> vref; vref < kernel_elf_base + (1 << pageBits) \<rbrakk> \<Longrightarrow>
+    table_index (pt_slot_offset max_pt_level riscv_global_pt_ptr vref) = elf_index"
+  using is_page_aligned_kernelELFPAddrBase
+  apply (simp add: table_index_riscv_global_pt_ptr elf_index_value toplevel_bits_value
+                   kernel_elf_base_def kernelELFBase_def)
+  apply (simp add: bit_simps add_ac)
+  apply (drule order_less_le_trans)
+   apply (rule word_plus_mono_right)
+    apply (rule kernelELF_plus_page[unfolded bit_simps toplevel_bits_value, simplified])
+   apply (simp add: pptrTop_def)
+  apply (simp add: bit_simps pptrTop_def mask_def is_aligned_nth)
   apply word_bitwise
-  apply (clarsimp simp: word_size)
+  apply clarsimp
   done
 
-lemma kernel_mapping_slots_0x1FE[simp]:
-  "0x1FE \<in> kernel_mapping_slots"
+lemma leq_elf_index:
+  "0x100 \<le> elf_index"
+  by (simp add: elf_index_value)
+
+lemma kernel_mapping_slots_elf_index[simp]:
+  "elf_index \<in> kernel_mapping_slots"
   by (simp add: kernel_mapping_slots_def pptr_base_def bit_simps pt_bits_left_def level_defs
-                pptrBase_def canonical_bit_def)
+                pptrBase_def canonical_bit_def leq_elf_index)
+
+lemma rewrite_eq_minus_to_plus_eq:
+  "(x = a - b) = (b + x = a)" for x :: "'a::ring"
+  by auto
+
+lemma merge_kernelELFPAddrBase_masks:
+  "((kernelELFPAddrBase && ~~ mask 30) +
+    (pptrTop + (kernelELFPAddrBase && mask 30) - kernelELFPAddrBase))
+   = pptrTop"
+  by (simp add: mask_out_sub_mask)
+
+lemma is_aligned_ptrFromPAddr_kernelELFPAddrBase:
+  "is_aligned
+     (ptrFromPAddr
+       (addr_from_ppn (ucast (kernelELFPAddrBase && ~~ mask toplevel_bits >> pageBits))))
+     (pt_bits_left max_pt_level)"
+  apply (simp add: addr_from_ppn_def ptrFromPAddr_def addrFromPPtr_def elf_index_value)
+  apply (simp add: pt_bits_left_def bit_simps level_defs ucast_ucast_mask toplevel_bits_value)
+  apply (simp add: pptrBase_def pptrBaseOffset_def paddrBase_def canonical_bit_def)
+  apply (rule is_aligned_add)
+   apply (metis and_not_mask is_aligned_andI1 is_aligned_shiftl_self shiftr_and_eq_shiftl)
+  apply (simp add: is_aligned_def)
+  done
+
+lemma kernelELFPAddrBase_addrFromKPPtr:
+  "\<lbrakk> kernel_elf_base \<le> vref; vref < kernel_elf_base + (1 << pageBits) \<rbrakk>
+   \<Longrightarrow> addr_from_ppn (ucast (kernelELFPAddrBase && ~~ mask toplevel_bits >> pageBits)) +
+       (vref && mask (pt_bits_left max_pt_level))
+       = addrFromKPPtr vref"
+  apply (simp add: addr_from_ppn_def ptrFromPAddr_def addrFromPPtr_def elf_index_value)
+  apply (simp add: pt_bits_left_def bit_simps level_defs ucast_ucast_mask toplevel_bits_value)
+  apply (simp add: addrFromKPPtr_def mask_shiftl_decompose flip: word_and_mask_shiftl)
+  apply (simp add: ac_simps neg_mask_combine)
+  apply (simp add: kernel_elf_base_def kernelELFBaseOffset_def)
+  apply (simp only: kernelELFBase_def)
+  apply (subst rewrite_eq_minus_to_plus_eq)
+  apply (simp add: add_ac merge_kernelELFPAddrBase_masks)
+  apply (subst word_plus_and_or_coroll)
+   apply (simp add: pptrTop_def mask_def)
+   apply word_bitwise
+  apply (subst (asm) word_plus_and_or_coroll)
+   apply (simp add: pptrTop_def mask_def)
+   apply word_bitwise
+  apply (drule order_less_le_trans)
+   apply (rule word_plus_mono_right)
+    apply (rule kernelELF_plus_page[unfolded bit_simps toplevel_bits_value, simplified])
+   apply (simp add: pptrTop_def)
+  apply (simp add: pptrTop_def mask_def)
+  apply word_bitwise
+  apply clarsimp
+  done
 
 lemma translate_address_kernel_elf_window:
-  "\<lbrakk> kernel_elf_base \<le> vref; vref < kernel_elf_base + (1 << 20) \<rbrakk> \<Longrightarrow>
+  "\<lbrakk> kernel_elf_base \<le> vref; vref < kernel_elf_base + (1 << pageBits) \<rbrakk> \<Longrightarrow>
    translate_address riscv_global_pt_ptr vref (ptes_of init_A_st) = Some (addrFromKPPtr vref)"
   apply (clarsimp simp: translate_address_def in_omonad pt_lookup_target_def
                         pt_lookup_slot_from_level_def)
   apply (simp add: ptes_of_init_A_st_global[THEN fun_cong] init_global_pt_def global_pte_def pte_ref_def)
-  apply (simp add: elf_window_1M is_aligned_pt_slot_offset_pte)
-  apply (simp add: bit_simps addr_from_ppn_def shiftl_shiftl)
-  apply (simp add: ptrFromPAddr_def addrFromPPtr_def)
-  apply (simp add: addrFromKPPtr_def kernelELFBaseOffset_def kernelELFPAddrBase_def kernelELFBase_def
-                   Kernel_Config.physBase_def mask_def)
-  apply (simp add: pt_bits_left_def bit_simps level_defs)
-  apply (rule conjI)
-   apply (simp add: pptrBase_def pptrBaseOffset_def paddrBase_def canonical_bit_def is_aligned_def)
-  apply (simp add: kernel_elf_base_def kernelELFBase_def kernelELFPAddrBase_def Kernel_Config.physBase_def)
-  apply (subst word_plus_and_or_coroll)
-   apply (simp add: canonical_bit_def word_size mask_def)
-   apply word_bitwise
-  apply (simp add: canonical_bit_def word_size mask_def)
-  apply word_bitwise
-  apply clarsimp
+  apply (simp add: elf_window_4k is_aligned_pt_slot_offset_pte elf_index_value)
+  apply (clarsimp simp: is_aligned_ptrFromPAddr_kernelELFPAddrBase kernelELFPAddrBase_addrFromKPPtr)
   done
 
 lemma kernel_window_init_st:
-  "kernel_window init_A_st = { pptr_base ..< pptr_base + (1 << 30) }"
-  by (auto simp: state_defs kernel_window_def)
+  "kernel_window init_A_st = { pptr_base ..< pptr_base + (1 << kernel_window_bits) }"
+  by (auto simp: state_defs kernel_window_def toplevel_bits_def bit_simps)
+
+lemma abs_kernel_window_sufficient:
+  "pptr_base + (1 << kernel_window_bits) \<le> kernel_elf_base"
+  unfolding pptr_base_def kernel_elf_base_def
+  using kernel_window_sufficient
+  by simp
+
+lemma abs_kernel_elf_window_at_least_page:
+  "kernel_elf_base + 2 ^ pageBits \<le> kdev_base"
+  unfolding kernel_elf_base_def kdev_base_def
+  using kernel_elf_window_at_least_page
+  by simp
+
+lemma kernel_elf_base_no_overflow:
+  "kernel_elf_base < kernel_elf_base + 2 ^ pageBits"
+  unfolding kernel_elf_base_def
+  by (rule kernelELFBase_no_overflow)
 
 lemma kernel_elf_window_init_st:
-  "kernel_elf_window init_A_st = { kernel_elf_base ..< kernel_elf_base + (1 << 20) }"
-  apply (clarsimp simp: state_defs kernel_elf_window_def kernel_elf_base_def kernelELFBase_def
-                        pptr_base_def pptrBase_def canonical_bit_def kernelELFPAddrBase_def
-                        Kernel_Config.physBase_def)
-  apply (rule set_eqI, clarsimp)
-  apply (rule iffI)
-   apply auto[1]
-  apply clarsimp
-  apply word_bitwise
-  apply clarsimp
-  done
+  "kernel_elf_window init_A_st = { kernel_elf_base ..< kernel_elf_base + (1 << pageBits) }"
+  using abs_kernel_window_sufficient
+  by (force simp: state_defs kernel_elf_window_def)
 
 lemma valid_global_vspace_mappings_init_A_st[simp]:
   "valid_global_vspace_mappings init_A_st"
@@ -322,29 +411,22 @@ lemma valid_global_vspace_mappings_init_A_st[simp]:
 lemma valid_uses_init_A_st[simp]: "valid_uses_2 init_vspace_uses"
 proof -
   note canonical_bit_def[simp]
-  have [simp]: "pptr_base < pptr_base + 0x40000000"
-    by (simp add: pptr_base_def pptrBase_def)
+  have [simplified, simp]: "pptr_base < pptr_base + (1 << kernel_window_bits)"
+    by (simp add: pptr_base_def pptrBase_def kernel_window_bits_def)
   have [simp]: "p \<le> canonical_user \<Longrightarrow> \<not> pptr_base \<le> p" for p
     by (rule notI, drule (1) order_trans)
        (simp add: canonical_user_def mask_def pptr_base_def pptrBase_def)
   have [simp]: "p \<le> canonical_user \<Longrightarrow> \<not> kernel_elf_base \<le> p" for p
-    by (rule notI, drule (1) order_trans)
-       (simp add: canonical_user_def mask_def kernel_elf_base_def kernelELFBase_def
-                  kernelELFPAddrBase_def Kernel_Config.physBase_def)
+    using canonical_user_kernel_elf_base by simp
   have [simp]: "p \<le> canonical_user \<Longrightarrow> \<not> kdev_base \<le> p" for p
     by (rule notI, drule (1) order_trans)
        (simp add: canonical_user_def mask_def kdev_base_def kdevBase_def)
-  have [simp]: "kernel_elf_base \<le> p \<Longrightarrow> \<not> p < pptr_base + 0x40000000" for p
-    by (rule notI, drule (1) order_le_less_trans)
-       (simp add: kernel_elf_base_def kernelELFBase_def pptr_base_def pptrBase_def
-                  kernelELFPAddrBase_def Kernel_Config.physBase_def mask_def)
-  have [simp]: "kdev_base \<le> p \<Longrightarrow> \<not> p < kernel_elf_base + 0x100000" for p
-    by (rule notI, drule (1) order_le_less_trans)
-       (simp add: kernel_elf_base_def kernelELFBase_def kdev_base_def kdevBase_def pptrBase_def
-                  kernelELFPAddrBase_def Kernel_Config.physBase_def mask_def)
-  have "pptr_base + 0x40000000 < kernel_elf_base + 0x100000"
-    by (simp add: kernel_elf_base_def kernelELFBase_def pptr_base_def pptrBase_def
-                  kernelELFPAddrBase_def Kernel_Config.physBase_def mask_def)
+  have [simplified, simp]: "kernel_elf_base \<le> p \<Longrightarrow> \<not> p < pptr_base + (1 << kernel_window_bits)" for p
+    using abs_kernel_window_sufficient by simp
+  have [simp]: "kdev_base \<le> p \<Longrightarrow> \<not> p < kernel_elf_base + 2 ^ pageBits" for p
+    using abs_kernel_elf_window_at_least_page by simp
+  have "pptr_base + (1 << kernel_window_bits) < kernel_elf_base + 2 ^ pageBits"
+    using abs_kernel_window_sufficient kernel_elf_base_no_overflow by simp
   thus ?thesis
     using canonical_user_pptr_base pptr_base_kernel_elf_base
     unfolding valid_uses_2_def init_vspace_uses_def window_defs
@@ -381,25 +463,42 @@ lemma valid_vspace_objs_init_A_st[simp]:
 lemma global_pt_kernel_window_init_arch_state[simp]:
   "obj_addrs init_global_pt riscv_global_pt_ptr \<subseteq>
      kernel_window_2 (riscv_kernel_vspace init_arch_state)"
-  apply (clarsimp simp: state_defs pptr_base_num bit_simps kernel_window_def kernel_elf_base_def)
+  apply (clarsimp simp: state_defs pptr_base_num bit_simps kernel_window_def kernel_elf_base_def
+                        kernel_window_bits_def)
   apply (rule conjI; unat_arith)
   done
 
 lemma idle_thread_in_kernel_window_init_arch_state[simp]:
   "{idle_thread_ptr..0x3FF + idle_thread_ptr} \<subseteq>
      kernel_window_2 (riscv_kernel_vspace init_arch_state)"
-  apply (clarsimp simp: state_defs pptr_base_num bit_simps kernel_window_def kernel_elf_base_def)
+  apply (clarsimp simp: state_defs pptr_base_num bit_simps kernel_window_def kernel_elf_base_def
+                        kernel_window_bits_def)
   apply (rule conjI; unat_arith)
   done
 
+lemma pptr_base_kernel_window_no_overflow:
+  "pptr_base \<le> pptr_base + (1 << kernel_window_bits)"
+  by (simp add: pptr_base_def pptrBase_def canonical_bit_def kernel_window_bits_def)
+
 lemma irq_node_pptr_base_kernel_elf_base:
-  "\<lbrakk>x \<le> pptr_base + (m + (mask cte_level_bits + 0x3000)); m \<le> mask (size irq) << cte_level_bits \<rbrakk>
-   \<Longrightarrow> \<not> kernel_elf_base \<le> x" for irq::irq
-  apply (simp add: word_size cte_level_bits_def mask_def pptr_base_def pptrBase_def
-                   kernel_elf_base_def kernelELFBase_def canonical_bit_def not_le
-                   kernelELFPAddrBase_def Kernel_Config.physBase_def)
-  apply unat_arith
-  done
+  fixes irq::irq
+  assumes "x \<le> pptr_base + (m + (mask cte_level_bits + 0x3000))"
+  assumes "m \<le> mask (size irq) << cte_level_bits"
+  shows "\<not> kernel_elf_base \<le> x"
+proof -
+  have less:
+    "\<And>m x. \<lbrakk> m \<le> x; x < 1 << kernel_window_bits \<rbrakk> \<Longrightarrow> pptr_base + m < kernel_elf_base"
+    using order_le_less_trans word_plus_strict_mono_right pptr_base_kernel_elf_base
+          pptr_base_kernel_window_no_overflow abs_kernel_window_sufficient
+    by fastforce
+  from assms show ?thesis
+    apply (simp add: not_le)
+    apply (erule order_le_less_trans)
+    apply (rule less[where x="mask cte_level_bits + 0x3000 + mask (size irq) << cte_level_bits"];
+           simp add: word_size cte_level_bits_def mask_def kernel_window_bits_def)
+    apply unat_arith
+    done
+qed
 
 lemma irq_node_in_kernel_window_init_arch_state':
   "\<lbrakk> init_irq_node_ptr + m \<le> x; x \<le> init_irq_node_ptr + m + mask cte_level_bits;
@@ -418,9 +517,10 @@ lemma irq_node_in_kernel_window_init_arch_state':
     apply (simp add: pptr_base_num canonical_bit_def is_aligned_def)
    apply (simp add: pptr_base_num cte_level_bits_def canonical_bit_def mask_def word_size)
    apply unat_arith
-  apply (simp add: kernel_elf_base_def kernelELFBase_def cte_level_bits_def canonical_bit_def
-                   mask_def init_irq_node_ptr_def pptr_base_num word_size kernelELFPAddrBase_def
-                   Kernel_Config.physBase_def)
+  apply clarsimp
+  apply (thin_tac "kernel_elf_base \<le> x \<longrightarrow> P" for x P)
+  apply (simp add: cte_level_bits_def canonical_bit_def mask_def init_irq_node_ptr_def
+                   pptr_base_num word_size kernel_window_bits_def)
   apply unat_arith
   apply clarsimp
   done
